const { jest, afterEach, afterAll, it, expect } = require('@jest/globals')

jest.mock('chalk', () => ({
	bold: (text) => text,
	red: (text) => text,
	bgRed: (text) => text,
}))

jest.spyOn(console, 'log').mockImplementation(() => { })
jest.spyOn(console, 'error').mockImplementation(() => { })

afterEach(() => {
	jest.clearAllMocks()
})

afterAll(() => {
	jest.restoreAllMocks()
})

const test = require('./test')

it('does not return false, given no failing test case', () => {
	const rules = {
		foo: {
			create(context) {
				return {
					Program(node) {
						if (node.body.length > 0) {
							context.report({
								node,
								message: 'bar'
							})
						}
					}
				}
			},
			tests: {
				valid: [{ code: '' }],
				invalid: [{ code: 'void(0)', errors: [{ message: 'bar' }] }],
			}
		}
	}

	expect(test(rules)).not.toBe(false)
	expect(console.log).toHaveBeenCalledWith(expect.stringMatching(/Done testing 1 rule/))
	expect(console.error).not.toHaveBeenCalled()
})

it('returns false, given a failing test case', () => {
	const rules = {
		foo: {
			create(context) {
				return {
					Program(node) {
						if (node.body.length > 0) {
							context.report({
								node,
								message: 'bar'
							})
						}
					}
				}
			},
			tests: {
				valid: [{ code: 'void(0)' }],
				invalid: [{ code: '', errors: [{ message: 'bar' }] }],
			}
		}
	}

	expect(test(rules)).toBe(false)
	expect(console.log).toHaveBeenCalledWith('ðŸ”´ foo')
	expect(console.error).toHaveBeenCalledWith(expect.stringMatching(/Should have no errors but had 1/))
})
